File: LeetCode - Top Interview 150/129.py
# 129. Sum Root to Leaf Numbers
# https://leetcode.com/problems/sum-root-to-leaf-numbers/


from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sumNumbers1(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        result = 0

        queue = []
        queue.append((root, str(root.val)))

        while queue:
            node, current_num = queue.pop(0)

            if node.left:
                queue.append((node.left, current_num + str(node.left.val)))
            if node.right:
                queue.append((node.right, current_num + str(node.right.val)))
            if not node.left and not node.right:
                result += int(current_num)

        return result

    def sumNumbers2(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        result = 0

        def traverse(node: Optional[TreeNode], prev_num: str):
            nonlocal result

            if not node:
                return

            current_num = prev_num + str(node.val)

            traverse(node.left, current_num)
            traverse(node.right, current_num)

            if not node.left and not node.right:
                result += int(current_num)

        traverse(root, "")

        return result

File: LeetCode - Algorithm/Binary Tree/129.ts
// 129. Sum Root to Leaf Numbers
// https://leetcode.com/problems/sum-root-to-leaf-numbers/

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function sumNumbers(root: TreeNode | null): number {
  let result = 0;

  function dfs(node: TreeNode, string: string = "") {
    const newString = string + String(node.val);

    if (!node.left && !node.right) {
      result += Number(newString);

      return;
    }

    if (node.left) dfs(node.left, newString);
    if (node.right) dfs(node.right, newString);
  }

  root && dfs(root);

  return result;
}

`
In this question, I could easily understand that it requires me to use DFS traversal.
`;
